File: llm_generated_code/ChatGPT/CPP/Level2/Q9.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

struct Item {
    int weight;
    int value;
    double valuePerWeight;
};

// Comparator function to sort items by value-to-weight ratio in descending order
bool compare(Item a, Item b) {
    return a.valuePerWeight > b.valuePerWeight;
}

double fractionalKnapsack(vector<int>& weights, vector<int>& values, int capacity) {
    int n = weights.size();
    vector<Item> items;

    // Create items with value-to-weight ratio
    for (int i = 0; i < n; i++) {
        items.push_back({weights[i], values[i], (double)values[i] / weights[i]});
    }

    // Sort items by value per weight
    sort(items.begin(), items.end(), compare);

    double totalValue = 0.0;

    for (const auto& item : items) {
        if (capacity >= item.weight) {  // Take whole item if capacity allows
            totalValue += item.value;
            capacity -= item.weight;
        } else {  // Take fraction of item if capacity is limited
            totalValue += item.valuePerWeight * capacity;
            break;
        }
    }

    return totalValue;
}

int main() {
    vector<int> weights = {2, 3, 5, 7};
    vector<int> values = {10, 5, 15, 7};
    int capacity = 8;

    double maxTreasureValue = fractionalKnapsack(weights, values, capacity);

    cout << "Maximum treasure value the hunter can carry: " << maxTreasureValue << endl;

    return 0;
}

File: llm_generated_code/ChatGPT/Python/Level2/Q9.py
def fractional_knapsack(weights, values, capacity):
    # Calculate value-to-weight ratio for each item and store them with their original indices
    items = sorted([(values[i] / weights[i], weights[i], values[i]) for i in range(len(weights))], reverse=True)

    total_value = 0.0
    for ratio, weight, value in items:
        if capacity >= weight:
            # Take the whole item if the remaining capacity allows
            capacity -= weight
            total_value += value
        else:
            # Take the fraction of the item that fits
            total_value += ratio * capacity
            break

    return total_value

# Example usage
weights = [2, 3, 5, 7]
values = [10, 5, 15, 7]
capacity = 8

max_value = fractional_knapsack(weights, values, capacity)
print("Maximum value the treasure hunter can carry:", max_value)

